Make more operators support `torch.complex32`

malfet opened the pull request:
Collection of changes aimed at allowing developer to create complex32 tensor,print it and access its elements
A followup after https://github.com/pytorch/pytorch/pull/43279

malfet committed:
Add kHalf time to `lt_cpu` kernel

All other comparison operations work for half floats other than `lt`

malfet committed:
Add support for kHalf to index_kernel, masked_fill_kernel and masked_select_kernel

malfet committed:
Fix typos

malfet committed:
Add `calc_erfinv` specialization for c10::Half

malfet committed:
Make erf, erfc and erfgamma templates accessible for c10::Half type

Do it by adding a dummy `[](T x) -> T { return f(x); }` lambda, which should be a no-op for base types

malfet committed:
Extend unary CPU operators to at::Half type

anjali411 commented on torch/csrc/utils/python_scalars.h:
```diff
@@ -40,8 +41,12 @@ inline PyObject* load_scalar(void* data, at::ScalarType scalarType) {
     case at::kFloat: return PyFloat_FromDouble(*(float*)data);
     case at::kDouble: return PyFloat_FromDouble(*(double*)data);
+    case at::kComplexHalf: {
+      auto data_ = (c10::complex<at::Half>*)data;
```
`auto data_ = reinterpret_cast<c10::complex<at::Half>*>(data);`

anjali411 commented on c10/util/Half.h:
```diff
@@ -374,6 +374,8 @@ struct alignas(4) complex<Half> {
   }
   inline complex(c10::complex<float> value)
       : real_(value.real()), imag_(value.imag()) {}
+  inline complex(c10::complex<double> value)
```
what is this change for?

malfet replied:
To support `torch::utils::store_scalar()` for c10::complex<Half>

anjali411 reviewed (changes requested):
please add tests to verify correctness of `torch.masked_select`, `torch.masked_fill` for BFloat16  and `torch.lt` for `Half`.

malfet committed:
Add ComplexHalf support to FillKernel

This makes it possible to call `torch.ones` and `torch.zeros` for `torch.complex32` tensors

malfet committed:
Add complex32 to test_view_as_real

malfet committed:
Stop special case treatment for torch.half in test_masked_fill

malfet commented:
@anjali411 do you want an explicit or implicit test? `print(torch.tensor([1+2j], dtype=torch.complex32))` implicitly tests all those operators. And can you please point me to the test that validates torch.lt for all dtypes?

malfet committed:
get_all_dtypes should include complex32 if appropriate option is passed

malfet committed:
Enable test_logical and test_masked_select for torch.half on CPU

This tests `torch.lt` for torch.half

anjali411 commented on torch/testing/__init__.py:
```diff
@@ -313,20 +313,25 @@ def all_types_and_complex_and(*dtypes):
 
-def get_all_complex_dtypes() -> List[torch.dtype]:
-    return [torch.complex64, torch.complex128]
+def get_all_complex_dtypes(include_complex32=False) -> List[torch.dtype]:
```
nit - `include_chalf=False` for a more consistent naming scheme. Also we should add the alias `chalf` for `complex32`

anjali411 commented on aten/src/ATen/cpu/vec256/vec256_base.h:
```diff
@@ -312,10 +312,10 @@ struct Vec256 {
   }
   Vec256<T> erf() const {
-    return map(std::erf);
+    return map([](T x) -> T { return std::erf(x); });
```
what is this and other similar changes for?

anjali411 reviewed (changes requested):
Currently this PR has some missing tests, for example, `torch.lt`, `torch.sqrt`, `index_put` etc. Please add tests for all ops enabled for `torch.half` and `torch.complex32`.

Also, I think we should add the alias `torch.chalf` for `torch.complex32` similar to numpy aliases `np.cfloat` for `np.complex64` and `np.cdouble` for np.complex128`. This can be done in a separate PR, but just added this note as an FYI.

Also we could add a section ComplexHalf support [here](https://github.com/pytorch/pytorch/issues/33152) to keep track of all the support being added and action items on the list. Would you like to go ahead and mention some action items on the issue?
